scaleway/compute_instance_helpers.go

```diff
@@ -0,0 +1,136 @@
+	userData := v.(map[string]interface{})
+	return hashcode.String(userData["key"].(string) + userData["value"].(string))
+}
+
+// orderVolumes return an ordered slice based on the volume map key "0", "1", "2",...
+func orderVolumes(v map[string]*instance.Volume) []*instance.Volume {
+	indexes := []string{}
+	for index := range v {
+		indexes = append(indexes, index)
+	}
+	sort.Strings(indexes)
+	var orderedVolumes []*instance.Volume
+	for _, index := range indexes {
+		orderedVolumes = append(orderedVolumes, v[index])
+	}
+	return orderedVolumes
+}
+
+// expandServerState converts the API state to terraform state or return an error.
+func expandServerState(fromState instance.ServerState) (string, error) {
+	switch fromState {
+	case instance.ServerStateStopped:
+		return ServerStateStopped, nil
+	case instance.ServerStateStoppedInPlace:
+		return ServerStateStandby, nil
+	case instance.ServerStateRunning:
+		return ServerStateStarted, nil
+	case instance.ServerStateLocked:
+		return "", fmt.Errorf("server is locked, please contact Scaleway support: https://console.scaleway.com/support/tickets")
+	}
+	return "", fmt.Errorf("server is in an invalid state, someone else might be executing action at the same time")
+}
+
+// computeServerStateToAction returns the action required to transit from a state to another.
+func computeServerStateToAction(previousState, nextState string, forceReboot bool) []instance.ServerAction {
+	if previousState == ServerStateStarted && nextState == ServerStateStarted && forceReboot {
+		return []instance.ServerAction{instance.ServerActionReboot}
+	}
+	transitionMap := map[[2]string][]instance.ServerAction{
+		{ServerStateStopped, ServerStateStopped}: {},
+		{ServerStateStopped, ServerStateStarted}: {instance.ServerActionPoweron},
+		{ServerStateStopped, ServerStateStandby}: {instance.ServerActionPoweron, instance.ServerActionStopInPlace},
+		{ServerStateStarted, ServerStateStopped}: {instance.ServerActionPoweroff},
+		{ServerStateStarted, ServerStateStarted}: {},
+		{ServerStateStarted, ServerStateStandby}: {instance.ServerActionStopInPlace},
+		{ServerStateStandby, ServerStateStopped}: {instance.ServerActionPoweroff},
+		{ServerStateStandby, ServerStateStarted}: {instance.ServerActionPoweron},
+		{ServerStateStandby, ServerStateStandby}: {},
+	}
+
+	return transitionMap[[2]string{previousState, nextState}]
+}
+
+// reachState executes server action(s) to reach the expected state
+func reachState(instanceApi *instance.API, zone utils.Zone, serverID, fromState, toState string, forceReboot bool) error {
+	for _, action := range computeServerStateToAction(fromState, toState, forceReboot) {
+		err := resource.Retry(ServerRetryFuncTimeout, func() *resource.RetryError {
+			err := instanceApi.ServerActionAndWait(&instance.ServerActionAndWaitRequest{
+				Zone:     zone,
+				ServerID: serverID,
+				Action:   action,
+				Timeout:  ServerWaitForTimeout,
+			})
+			if isSDKError(err, "expected state [\\w]+ but found [\\w]+") {
```

Comment:
I really don't like regex error checking 😢.

Reply:
me too, do you see a better solution?

Reply:
either a specific type or an error constant in the SDK like `ErrRegionNotFound` in terraform